pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/util/ValueReaderComparisons.java

```diff
@@ -0,0 +1,174 @@
+                return i + ((littleEndian
+                        ? Long.numberOfTrailingZeros(difference)
+                        : Long.numberOfLeadingZeros(difference)) >>> 3);
+            }
+        }
+        for (; i < limit; i++) {
+            byte ours = dataBuffer.getByte(startOffset + i);
+            byte theirs = buffer.get(i);
+            if (ours != theirs) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    static int compareUtf8Bytes(PinotDataBuffer dataBuffer, long startOffset, int length, boolean padded,
+                           byte[] bytes) {
+        // can use MethodHandles.byteArrayViewVarHandle here after dropping JDK8
+        ByteBuffer buffer = ByteBuffer.wrap(bytes);
+        int mismatchPosition = mismatch(dataBuffer, startOffset, length, buffer);
+        if (mismatchPosition == -1) {
+            // need to figure out whether the unpadded string is longer than the parameter or not
+            if (padded && bytes.length < length) {
+                if (bytes.length == 0) {
+                    // just need nonzero first byte
+                    return dataBuffer.getByte(startOffset) == 0 ? 0 : 1;
+                } else if (bytes[bytes.length - 1] == 0) {
+                    // we can't store zero padding, but can have it in parameters, check the last byte
+                    return dataBuffer.getByte(startOffset + bytes.length - 1) == 0 ? -1 : 0;
+                } else {
+                    // check if the stored string continues beyond the length of the parameter
+                    return dataBuffer.getByte(startOffset + bytes.length) == 0 ? 0 : 1;
+                }
+            } else {
+                // then we know the length precisely or know that the parameter is at least as long as we can store
+                return length - bytes.length;
+            }
+        }
+        byte ourByte = dataBuffer.getByte(startOffset + mismatchPosition);
+        if (ourByte == 0) {
+            // then we are storing the empty string
+            return -1;
+        }
+        byte theirByte = bytes[mismatchPosition];
+        if (theirByte == 0) {
+            return 1;
+        }
+        // we know the position of the mismatch but need to do utf8 decoding before comparison
+        // to respect collation rules
+        return compareUtf8(dataBuffer, startOffset, buffer, mismatchPosition);
+    }
+
+    private static int compareUtf8(PinotDataBuffer ourBuffer, long ourStartOffset,
+                           ByteBuffer theirBuffer, int mismatchPosition) {
+        char ours1 = '?';
+        char ours2 = '?';
+        char theirs1 = '?';
+        char theirs2 = '?';
+
+        // 1. seek backwards from mismatch position in each buffer to find start of each utf8 sequence
+        long ourPosition = ourStartOffset + mismatchPosition;
+        while (ourPosition > ourStartOffset && isUtf8Continuation(ourBuffer.getByte(ourPosition))) {
```

Comment:
The boundary check is redundant. We can also move both our and their position together for better performance

Reply:
Why do you think it's redundant? We could potentially go backwards forever if the UTF8 is corrupt. Assuming valid UTF8, the mismatch sequences will either have the same length (e.g. 4 bytes starting with 0xFX, with a mismatch in any of the next 3 bytes, or mismatching ASCII characters), so the loops could be fused, but this would be less resilient to corrupt data. The maximum length for either of these loops is 4 iterations, and they access different buffers, so the performance impact that can be made here is minimal.

Reply:
I've changed this so there's only one loop, having used the `String` class before storage and in the query layer makes this safe enough.